### ISSUE TITLE
Completion algorithm: try prefix, fallback to fuzzy if, and only if, prefix gives no results

### ISSUE THREAD

**Issue by k0ral** [nushell/nushell#11211]
### Related problem

The `completions.algorithm` option currently supports 2 values: `fuzzy` and `prefix`.

What about supporting a 3rd mode for {what the title says} ?

### Describe the solution you'd like

- `completions.algorithm` can be set to `prefix_fallback_fuzzy` (yeah, naming things is hard...) 
- when set to that mode, auto-completion works like this:
  - first, look for prefix matches ; if there are some, act as if `completions.algorithm` were set to `prefix`
  - if there are no prefix matches, act as if `completions.algorithm` were set to `fuzzy`   

### Describe alternatives you've considered

I'm not aware of any trick to emulate that behavior with current nushell.

**Comment by fdncred** [comment]
Hmmm, I'd have to try it out to see if it fits. Seems like it would be ok, but I just don't know. I could see them named `prefix`, `fuzzy`, or `fallback` and fallback means its set to prefix, falling back to fuzzy.

**Comment by curldivergence** [comment]
Hi! Sorry for hijacking the thread (I found it by keyword "fuzzy"), but it didn't seem worth it to create a new issue.
So I have a question: what exactly is the difference between "prefix" and "fuzzy"? E.g. in `fish` I can type any substring, e.g. `argo`, and press Arrow-Up, and it will start scrolling over past commands that contained it (e.g. `cargo run`, `which cargo` etc.). But unfortunately, I can't get the same behavior in `nu` whatever I do: at first I expected that this `fuzzy` option will enable this behavior, but it did not. Then I gave up and enabled `fish` completer as an external one, but it didn't work either :( Thanks!

**Comment by curldivergence** [comment]
@fdncred Thanks a lot for looking into this! A couple more questions though, if you don't mind :)
- I can see there is `SubstringSearch` enum variant there, so my example with `argo` would still be expected to work, wouldn't it?
- could you elaborate on the difference between this `SubstringSearch` and fuzzy search, please? The fuzzy one would be expected to find other history items within a given Levenshtein distance, or I understand it wrong?
- and generally speaking, this behavior (scrolling through history items after typing something in the shell and pressing Arrow-Up) logically belongs to history or to completions?
Thank you!

**Comment by fdncred** [comment]
I'm guessing all of this is a reedline thing vs a nushell thing but could be wrong.

1. I think so? i haven't messed with this much.
2. substring maybe should work like `argo`. fuzzy would work like typing `gco` and matching `git checkout`
3. i think what you're looking for is history vs completions because you don't want to scroll through completions, you want to see the next history result that matches what you type.

**Comment by curldivergence** [comment]
@fdncred Got it, thanks a lot for you help!

**Comment by tobiasBora** [comment]
Out of curiousity, why don't we allow the user to write custom completion algorithm?

**Comment by unnamedname** [comment]
Indeed, another [nice thing to have from Fish](https://github.com/nushell/nushell/issues/7640). 

Users’ expectations regarding completion for different parts of the command line are rarely one-size-fits-all. For me, the problem with the current `completions.algorithm` is that `fuzzy` often results in redundant command/argument suggestions, whilst `prefix`, conversely, leads to no results for (fallback) path completions. Furthermore, for up-arrow key completion (from command history), Fish’s substring matching is more intuitive (than Nushell’s prefix-only approach).

**Comment by flying-sheep** [comment]
Fuzzy can be quite confusing, but I guess that’s only when things are broken.

E.g. I use the `fish` completer for `git` but failed to install `fish` on a system. I also have

```nushell
export extern "yarn plugin import from sources" [...] {...}
```

So completing `git ` with “fuzzy” gave me `yarn plugin import from sources ` lol.

But maybe that just means that fuzzy isn’t for me as it kinda always gives something.

**Comment by j-chmielewski** [comment]
@fdncred I think merging https://github.com/nushell/nushell/pull/18761 closed this issue by mistake - it doesn't seem to me that it implements the "prefix-to-fuzzy" completion algorithm.

I would be interested in implementing this behavior. Please let me know if you think this would be a welcome contribution and perhaps reopen the issue?

**Comment by fdncred** [comment]
@j-chmielewski ya, have at it?
@philocalyst thoughts on the comment just above? Was this issue closed by mistake?
